spring-integration-file/src/main/java/org/springframework/integration/file/remote/AbstractLineByLineRemoteFileMessageSource.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.file.remote;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Comparator;
+
+import org.springframework.integration.file.remote.session.Session;
+import org.springframework.integration.support.AbstractIntegrationMessageBuilder;
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+
+/**
+ * An {@link AbstractRemoteFileStreamingMessageSource} that emits individual
+ * lines.
+ *
+ * @author Gary Russell
+ * @since 4.3
+ *
+ */
+public abstract class AbstractLineByLineRemoteFileMessageSource<F> extends AbstractRemoteFileStreamingMessageSource<F> {
```

Comment:
Consider a very large file; with the current solution we have to download the whole file, then start splitting it.

With this technique, we don't have to write to disk, and we emit messages directly from the input stream.